nixos/modules/services/games/asf.nix

```diff
@@ -0,0 +1,200 @@
+          };
+        };
+      };
+      default = {
+        enable = true;
+        package = pkgs.ASF-ui;
+      };
+      example = {
+        enable = false;
+      };
+      description = "The Web-UI hosted on 127.0.0.1:1242.";
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.ArchiSteamFarm;
+      description =
+        "Package to use. Should always be the latest version, for security reasons and since this module uses very new features.";
+    };
+
+    dataDir = mkOption {
+      type = types.path;
+      default = "/var/lib/asf";
+      description = "State directory of Archis-Steam-Farm";
+    };
+
+    config = mkOption {
+      type = types.attrs;
+      description = ''
+        The ASF.json file, all the options are documented here: https://github.com/JustArchiNET/ArchiSteamFarm/wiki/Configuration#global-config.
+        Do note that AutoRestart is automatically set to false, since nix takes care of updating everything.
+        UpdateChannel is also set to 0 for the same reason.
+        Headless is also set to true, since there is no way to provide input via a systemd service.
+        You should try to keep ASF up to date since upstream does not provide support for anything but hte latest version and you're exposing yourself to all kinds of issues - as is outlined here: https://github.com/JustArchiNET/ArchiSteamFarm/wiki/Configuration#updateperiod
+      '';
+      example = {
+        Statistics = false;
+      };
+      default = { };
+    };
+    bots = mkOption {
+      type = types.attrsOf (types.submodule {
+        options = {
+          username = mkOption {
+            type = types.str;
+            description =
+              "Name of the user to log in. Default is attribute name.";
+            default = "";
+          };
+          passwordFile = mkOption {
+            type = types.path;
+            description =
+              "Path to a file containig the password. The file must be readable by the <literal>asf</literal> user/group.";
+          };
+          enabled = mkOption {
+            type = types.bool;
+            default = true;
+            description = "Whether to enable the bot on startup.";
+          };
+          settings = mkOption {
+            type = types.attrs;
+            description =
+              "Additional settings that are documented here: https://github.com/JustArchiNET/ArchiSteamFarm/wiki/Configuration#bot-config";
+            default = { };
+          };
+        };
+      });
+      description = ''
+        Bots name and configuration
+      '';
+      example = {
+        exampleBot = {
+          username = "alice";
+          passwordFile = "/my/super/secret/password";
+          settings = { SteamParentalCode = "1234"; };
+        };
+      };
+      default = { };
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    users = {
+      users.asf = {
+        home = cfg.dataDir;
+        createHome = true;
```

Comment:
`createHome` is usually a bad idea. See https://github.com/NixOS/nixpkgs/pull/115332#issuecomment-792287827 for some examples of why. Consider https://github.com/NixOS/nixpkgs/pull/93001 as an example of a better way to provision the data directory.

Reply:
So I remove the createHome, and instead pass StateDirectory to the service? As far as I can tell it wants a literal name of the dir, not a path. Is there some way to programmatically get the name without `dirname` (which i don't have at build-time)?

Reply:
> So I remove the createHome, and instead pass StateDirectory to the service?

Yes, that is what I suggest.

Creating the directory conditional on the user using the default value is a good move. You don't need to infer anything - just make assumptions and document that the sysadmin needs to provision the directory themselves if they want a different directory:

```nix
    systemd.services = {
      asf = {
        # ...
        serviceConfig = lib.mkMerge [
          {
            User = "asf";
            Group = "asf";
            WorkingDirectory = cfg.dataDir;
            Type = "simple";
            # ...
          }
          (lib.mkIf (cfg.dataDir == "/var/lib/asf") { StateDirectory = "asf"; })
        ];
      };
    };
```

You can set the `description` of `dataDir` to something like this:
> The ASF home directory used to store all repositories. If left as the default value
this directory will automatically be created before the ASF server starts, otherwise
the sysadmin is responsible for ensuring the directory exists with appropriate ownership
and permissions.

Reply:
I really dislike this, what's the purpose of a declarative configuration if you have to create an directory on your own... But you're right, there's no other way.

Reply:
`systemd` is **very** declarative - `StateDirectory` is very declarative! You shouldn't hate it! From observations I have noticed that **many** of the problems that come up, when using anything other than `StateDirectory` (ie. `createHome` or `systemd.tmpfiles`), are because we're **forcing** a declarative answer on the sysadmins stateful solution. Network attached storage cannot be made stateless (well... not easily at least, and not in a single module :wink:) so we need to resist trying to make everything fit our paradigm. Give the sysadmin an escape hatch.

I hope this makes you hate the solution less.